include/grpc++/impl/codegen/call.h

```diff
@@ -285,33 +284,64 @@ class CallOpSendMessage {
 
  protected:
   void AddOp(grpc_op* ops, size_t* nops) {
-    if (send_buf_ == nullptr) return;
+    if (!send_buf_.Valid()) return;
     grpc_op* op = &ops[(*nops)++];
     op->op = GRPC_OP_SEND_MESSAGE;
     op->flags = write_options_.flags();
     op->reserved = NULL;
-    op->data.send_message.send_message = send_buf_;
+    op->data.send_message.send_message = send_buf_.c_buffer();
     // Flags are per-message: clear them after use.
     write_options_.Clear();
   }
-  void FinishOp(bool* status) {
-    g_core_codegen_interface->grpc_byte_buffer_destroy(send_buf_);
-    send_buf_ = nullptr;
-  }
+  void FinishOp(bool* status) { send_buf_.Clear(); }
 
  private:
-  grpc_byte_buffer* send_buf_;
+  template <class M, class T = void>
+  class MessageSerializer;
+
+  ByteBuffer send_buf_;
   WriteOptions write_options_;
 };
 
 template <class M>
+class CallOpSendMessage::MessageSerializer<
+    M, typename std::enable_if<std::is_same<
```

Comment:
Could this be simplified a little with:

```c++
template <class T> T example(); // no need to provide an implementation
```

So that the enable_if becomes:
```c++
std::enable_if<std::is_same<::grpc::Status, decltype(SerializationTraits<M>::Serialize(
  example<M>(), example<grpc_byte_buffer**>(), example<bool*>()))>::value>::type>
```

I feel like the static casts are obscuring intent here

Reply:
Yeah, I'll give that a try